src/accapi/client.py

```diff
@@ -276,15 +266,27 @@ def request_hud_page(self, pageName: str):
         )
 
     def _run(self):
-        while not self._stopSignal:
-            try:
-                messageType = self._receive("B")
-            except ConnectionResetError:
-                self._update_connection_state("lost")
-                break
-            self._receiveMethods[messageType]()
-        self._socket.close()
-        self._socket = None
+        self._request_connection()
+        try:
+            while not self._stopSignal:
+                try:
+                    message = self._socket.recv(self._udpBufferSize)
+                except socket.timeout:
+                    self._update_connection_state("timeout")
+                    continue
+                except ConnectionResetError:
+                    self._update_connection_state("lost")
+                    break
+                # If we received a message after a timeout and we were already registered, change connection state
+                if self._connectionState == "timeout" and self._connectionId is not None:
+                    self._update_connection_state("established")
+                messageType = struct.unpack("B", message[:1])[0]
+                self._receiveMethods[messageType](list(message[1:]))
+        finally:
+            if self._connectionId is not None:
```

Comment:
There are more things than just a missing connection ID that can go wrong here, in #4, this is now a try/except inside of the finally.